File: rust/parity/specs/sbom_validator.py
"""Parity spec for PiSBOMValidator.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiSBOMValidator"

_mod = load_py_agent("pi_sbom_validator.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiSBOMValidator()
    out = agent.validate_sbom(_mod.SBOMInput(**data))
    return out.model_dump()


# Clean SBOM: has a signature block, no banned/risky licenses -> PASSED.
_CLEAN = '{"bomFormat": "CycloneDX", "components": [], "signature": {"algorithm": "RS256"}}'
# Banned AGPL copyleft (risk 85) but signature present -> only license issue.
_AGPL = '{"components": [{"license": "AGPL-3.0"}], "attestation": "present"}'
# Risky GPL-3.0 (risk 50) but attestation present -> only license issue, WARN/FAILED.
_GPL3 = '{"components": [{"license": "GPL-3.0"}], "signature": "0xdeadbeef"}'
# GPLv3 spelling variant with attestation present.
_GPLV3 = 'PackageLicenseDeclared: GPLv3\nSignatureBlock: attestation here'
# No banned license but missing signature/attestation (risk 60).
_NOSIG = '{"components": [{"name": "left-pad", "version": "1.0.0"}]}'
# AGPL AND missing signature: both gates trip, risk maxes at 85.
_AGPL_NOSIG = '{"components": [{"license": "AGPL-3.0"}]}'
# Uppercase content to exercise .lower() normalization.
_UPPER = '{"COMPONENTS": [{"LICENSE": "GPL-3.0"}], "SIGNATURE": "ABC"}'

SAMPLES = [
    # Clean / passing input.
    {"input": {"sbom_path": "a.json", "sbom_content": _CLEAN, "format": "cyclonedx"}},
    # Banned AGPL copyleft license path.
    {"input": {"sbom_path": "a.json", "sbom_content": _AGPL, "format": "cyclonedx"}},
    # Risky GPL-3.0 license path.
    {"input": {"sbom_path": "a.json", "sbom_content": _GPL3, "format": "cyclonedx"}},
    # GPLv3 spelling variant.
    {"input": {"sbom_path": "a.spdx", "sbom_content": _GPLV3, "format": "spdx"}},
    # Missing signature/attestation path.
    {"input": {"sbom_path": "a.json", "sbom_content": _NOSIG, "format": "cyclonedx"}},
    # Both license + attestation gates trip.
    {"input": {"sbom_path": "a.json", "sbom_content": _AGPL_NOSIG, "format": "cyclonedx"}},
    # Uppercase content (case-insensitivity).
    {"input": {"sbom_path": "a.json", "sbom_content": _UPPER, "format": "cyclonedx"}},
    # Empty content edge case (no license, missing signature -> risk 60).
    {"input": {"sbom_path": "a.json", "sbom_content": "", "format": "cyclonedx"}},
    # Non-strict env -> risk>30 no longer fails; WARN path with is_secure True.
    {"input": {"sbom_path": "a.json", "sbom_content": _AGPL, "format": "cyclonedx"},
     "env": {"PI_SBOM_STRICT_MODE": "false"}},
    # Strict env explicit -> FAILED path.
    {"input": {"sbom_path": "a.json", "sbom_content": _AGPL, "format": "cyclonedx"},
     "env": {"PI_SBOM_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/sbom_validator.rs
//! Port of `pi_micro_agents/pi_sbom_validator.py`.
//!
//! Validates SPDX/CycloneDX SBOMs for license compliance and missing
//! signatures/attestations. Behaviour is a line-for-line mirror of the Python
//! original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub sbom_path: String,
    pub sbom_content: String,
    pub format: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub license_issues: Vec<String>,
    pub missing_attestations: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true". When the var is unset, defaults to true.
fn is_strict_mode() -> bool {
    match std::env::var("PI_SBOM_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn validate_sbom(input: &Input) -> Output {
    let content = input.sbom_content.to_lowercase();
    let mut license_issues: Vec<String> = Vec::new();
    let mut missing_attestations: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // Check for banned licenses (copyleft licenses that violate standard enterprise compliance rules)
    if content.contains("agpl") || content.contains("agpl-3.0") {
        license_issues.push("Banned Copyleft License: AGPL-3.0 detected in dependency tree.".to_string());
        risk_score = risk_score.max(85.0);
    } else if content.contains("gpl-3.0") || content.contains("gplv3") {
        license_issues.push("Risky Copyleft License: GPL-3.0 detected in dependency tree.".to_string());
        risk_score = risk_score.max(50.0);
    }

    // Check for missing signature / attestation patterns
    if !content.contains("signature") && !content.contains("attestation") {
        missing_attestations
            .push("Missing Cryptographic Signature: No attestation blocks found in SBOM.".to_string());
        risk_score = risk_score.max(60.0);
    }

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec {
        "PASSED".to_string()
    } else {
        "FAILED_SBOM_VALIDATION".to_string()
    };
    if risk_score > 0.0 && is_sec {
        status = "WARN_SBOM_VALIDATION".to_string();
    }

    Output {
        is_secure: is_sec,
        license_issues,
        missing_attestations,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = validate_sbom(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(content: &str) -> Output {
        validate_sbom(&Input {
            sbom_path: "sbom.json".into(),
            sbom_content: content.into(),
            format: "cyclonedx".into(),
        })
    }

    #[test]
    fn clean_sbom_passes() {
        // has a signature, no banned licenses -> no risk, PASSED
        let o = run("{\"components\": [], \"signature\": \"abc\"}");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.license_issues.is_empty());
        assert!(o.missing_attestations.is_empty());
    }

    #[test]
    fn agpl_banned_fails_in_strict() {
        // contains AGPL and a signature -> only license issue, risk 85, FAILED in strict
        let o = run("license: AGPL-3.0 with signature present");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 85.0);
        assert_eq!(o.status, "FAILED_SBOM_VALIDATION");
        assert_eq!(
            o.license_issues,
            vec!["Banned Copyleft License: AGPL-3.0 detected in dependency tree."]
        );
        assert!(o.missing_attestations.is_empty());
    }

    #[test]
    fn gpl3_risky_with_missing_attestation() {
        // GPL-3.0 (50) and no signature/attestation (60) -> max 60, FAILED in strict
        let o = run("dependency uses GPL-3.0 license");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 60.0);
        assert_eq!(o.status, "FAILED_SBOM_VALIDATION");
        assert_eq!(
            o.license_issues,
            vec!["Risky Copyleft License: GPL-3.0 detected in dependency tree."]
        );
        assert_eq!(
            o.missing_attestations,
            vec!["Missing Cryptographic Signature: No attestation blocks found in SBOM."]
        );
    }
}
